Examples to show the mixers new configuration model.

guptasu opened the pull request:
Samples to show the new Mixer model where:
The design document for the new Model is here:
https://docs.google.com/document/d/1elqeOTDCGSaKP7GI0mrNkZDtt0dq8TbarEt0jUNTk-M/edit

Sample templates are under : configs/templates/.
Sample generated Interfaces that adapters would implement are under configs/templates/generated_interfaces_for_adapters/**/
Sample yaml configurations written by the Operators are under configs/sample_yamls

guptasu committed:
New model list checker sample (#6)

* Update mixer docs location (#750)

* Initial commit to show list checker configuration using the new mixer config model

* Some renaming and using an annotation to explicitly state fields that would come from `Constructor`

* Add alternative template style using nested messages.

* Addded namespace to the types and templates.

* Address comments from mtail and removed namespacing to avoid confusion for now.

ZackButcher committed:
log template codegen (#9)

* first pass at what codegen might look like

* include an example config file too

douglas-reid committed:
Add metric generation example to proposal (#8)

mandarjog committed:
New config quota samples (#7)

* Update mixer docs location (#750)

* quota experiment

* add template reference

guptasu committed:
Consolidated all the examples to conform to one format.

guptasu committed:
Merge branch 'master' into UnifySamples

guptasu committed:
cleanup for PR

guptasu committed:
rename the executable to txt

guptasu committed:
Address Martin's comments.

guptasu committed:
Removed the templateName from the processor interface methods.

guptasu committed:
Rename params to a oneoff fields in the service config yaml.

guptasu committed:
Make handlers use oneof params

louiscryan commented on configs/sample_yamls/serviceops/sample_service_config.yml:
```diff
@@ -0,0 +1,172 @@
+#
+########################## types ########################
+#
+types:
```
As discussed with @mandarjog I don't think these 'types' are adding any value here. They are 1:1 with constructors and that correspondence is likely to hold in more realistic configurations.

geeknoid replied:
The types define the shape of the data we will send to the backends. They are strictly necessary for the model to function end-to-end.

What we can support is some form of type-inference based on the defined constructors. We've discussed this, agreed it was a reasonable idea, and decided to proceed without them for the initial implementation. Once everything is working end to end, we can evaluate whether inference would deliver good bang for the buck and then we can add it then.

louiscryan replied:
It seems we're all in agreement that type inference is possible from constructors without a loss of fidelity.

Just to be clear you're advocating delaying doing that work until you've gotten other things working? That's reasonable although I do have concerns about the user-experience of authoring config which is what I was trying to focus on in my comments. 

Its pretty clear that there's some collateral damage to the samples in this PR because of naming confusion (see my other comments). This seems to be caused by the semantic overlap of constructor and type. I wouldn't want to live with this in the  product if it can be avoided.

geeknoid replied:
Yes, given that types are architecturally necessary, we will proceed to implementing the model where types are explicitly listed by the operator as stand-alone concepts. Once we have it working, then we will figure out the appropriate syntax to express type inference.

I think we can improve the name of entities in the example to improve the situation without having to change anything else.

Once again, I believe there is an upper limit to the conceptual complexity and total size that is reasonable for people to author against in YAML. We will need a bona fide UI to make all this usable.

louiscryan replied:
I'm OK with deferring as long as:
- we have a tracking bug to make sure we revisit this 
- we can show that the config is legible with types in place

Given that Beta should achieve API/config stability this must be resolved before that release

guptasu replied:
Captured in issue https://github.com/istio/mixer/issues/802

Thanks Louis for adding the issue.

louiscryan commented on configs/sample_yamls/serviceops/sample_service_config.yml:
```diff
@@ -0,0 +1,172 @@
+types:
+# QUOTA
+- name: ratelimit_by_service_by_consumer
+  quota_params:
```
is correspondence with template type inferred by parsing 'xxx_params' ?

Seems like it will lead to readability issues. Wouldn't it be better to nest each entry under their associated template type name

e.g.
```
constructors:
  quota:
    name:
    params:
...
  list_check:
    name:
    params:
...
```

This comment applies equally to constructors as it does to types.

guptasu replied:
@mandarjog @geeknoid : 

The plan is to create a oneof from all the possible templates, during the mixer build time, that would be part of the types array
message ServiceConfig {
  repeated Type types = 1;
  ..
}
message Type {
  string name = 1
  oneof Params {
    foo.bar.Metric metric_param = 2;
    foo.bar.Qupta quota_param = 3;
  }
}

@louiscryan  you are proposing we do something like:

message ServiceConfig {
  repeated QuotaType quota_types = 1;
  repeated MetricType metric_types = 2;
  ..
}
message QuotaType {
  string name = 1;
  // We don't need param's then. We can just flatten the fields of param's here
  map<string, istio.mixer.v1.config.descriptor.ValueType> dimensions = 2;

  int64 max_amount = 5;
  google.protobuf.Duration expiration = 6;
}
Thoughts from Mixers ?

geeknoid replied:
I think there is value in grouping all these things under "types" and "constructors". Otherwise, it's not clear what you're looking at in config.

louiscryan replied:
Fixed my indenting above, wasn't suggesting we get rid of the outer 'types' / 'constructors' but I was suggesting that sequences of types are kept together.

Not a strong opinion btw, just want to make sure we have some natural readability

guptasu replied:
I think the idea of hoisting the template types and template constructors above from the params is a good idea.
I have added a issue to address this https://github.com/istio/mixer/issues/803

louiscryan commented on configs/sample_yamls/serviceops/sample_service_config.yml:
```diff
@@ -0,0 +1,172 @@
+      consumer: STRING
+    maxAmount: 500
+    expiration: 1s
+- name: ratelimit_by_service_by_datacenter
```
note that these names are semantic (service & datacenter) and are more applicable to a 'constructor' than a 'type'.

guptasu replied:
Type and Constructors both can have the same name if operator choose to do so. Not sure what would be a good name for a Type here. @mandarjog any ideas ?

louiscryan commented on configs/sample_yamls/serviceops/sample_service_config.yml:
```diff
@@ -0,0 +1,172 @@
+- name: response_duration
+  type: istio_time_distribution
+  metric_params:
+    value: response.latency | response.duration | "0ms"
```
Are time literals quoted or not? Inconsistent usage

guptasu replied:
I think quoted and unquoted both should work. Removed the quota

louiscryan commented on configs/sample_yamls/serviceops/sample_service_config.yml:
```diff
@@ -0,0 +1,172 @@
+#
+contructors:
+# QUOTA
+- name: rl_by_service_by_consumer
```
.... and the constructors now have weird names because the 'types' consumed the natural names...

guptasu replied:
Will fix once the name on the type is fixed. TODO

guptasu committed:
Removed quotes from duration.

guptasu commented:
@louiscryan Based on the review so far, there are two issues https://github.com/istio/mixer/issues/803 and https://github.com/istio/mixer/issues/802 that came out of it.

Also, based on the above conversations, it seems we are okay to get back to these issues once we have an end to end thing working. Make sure we address them before Beta.

Can you approve if you think rest of the stuff in this example PR looks fine ?

+mtail @geeknoid

louiscryan commented:
Other than those two I'm happy

On Thu, Jun 8, 2017 at 11:49 AM, guptasu <notifications@github.com> wrote:

> @louiscryan <https://github.com/louiscryan> Based on the review so far,
> there are two issues istio#803 <https://github.com/istio/mixer/issues/803>
> and istio#802 <https://github.com/istio/mixer/issues/802> that came out
> of it.
>
> Also, based on the above conversations, it seems we are okay to get back
> to these issues once we have an end to end thing working. Make sure we
> address them before Beta.
>
> Can you approve if you think rest of the stuff in this example PR looks
> fine ?
>
> +mtail @geeknoid <https://github.com/geeknoid>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/guptasu/mixer/pull/13#issuecomment-307193776>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AIoKPNMY1qUSP9tY32G_fABWUhSYOOJaks5sCEI_gaJpZM4NxwOb>
> .
>

guptasu committed:
Updated the examples based on the discussion.

Does not include the type inference example since there are still some unknowns.

Signed-off-by: guptasu <guptasu@google.com>

geeknoid commented on configs/sample_yamls/meshops/handlers.yml:
```diff
@@ -0,0 +1,21 @@
+handlers:
+- name: versionList
+  adapter: genericListChecker
```
Given generalListChecker_params below, do we need this adapter field? Seems redundant.

@mandarjog @ZackButcher @douglas-reid

geeknoid replied:
Removing adapter would be consistent with remove template from the types below

guptasu replied:
@mandarjog proposed we can get rid of the oneof and keep the google.protobuf.struct. The config server will do the conversion to bytes and save in DB. When mixer gets it back it will use the adapter name  to do the correct loading of bytes to adapters specific proto object. Removing the reference to adapter for now since it is redundant with the oneof field that we are using here 'generalListChecker_params'. Mandar's suggestion was in context of params for Constructors and Types, but the same argument holds for adapter params too.

geeknoid commented on configs/templates/ListTemplate.proto:
```diff
@@ -0,0 +1,18 @@
+
+
+message Type {
+  // EMPTY, since nothing type needs to specify for the
```
Incomplete comment.

guptasu replied:
DONE

geeknoid commented on configs/templates/LogTemplate.proto:
```diff
@@ -0,0 +1,56 @@
+// COPY OF https://github.com/istio/api/blob/master/mixer/v1/config/descriptor/log_entry_descriptor.proto
```
Shoudl we remove this comment? This is 'inspired by' but is definitely not a copy.

guptasu replied:
DONE

guptasu committed:
fixed typos and removed redundant adapter string from handlers.

guptasu reviewed (commented):
@geeknoid @mandarjog there are some pending comment, PTAL.

geeknoid reviewed (approved):
This looks good.

Let's not check this in though, I don't think we need this in the source base. Soon, we'll have the real deal. But in the meantime, this will serve as reference and something we can show folks that want to see where we're headed.

Thanks for putting this together.

guptasu commented:
@geeknoid, I also prefer to not checkin since it is easy to comment on this live sample PR. We can update it as we learn more. 
Anyways, the checkin would be in My Repo 'guptasu' so, submitting it won't help anyone as it will become deal branch from outside world. 

BTW, do you want to run this by Sven, while you are there in MTV ?

guptasu committed:
Unified example3 (#18)

Update sample for Mixer's configuration.

Major highlights:

From the template, Type message is gone, From the Yaml file, types: is gone.
Along with the Go interface .go file, we are now generating a proto file for each template proto given by user. This proto file contains the *Param message (stringified Constructor, example MetricParam) and the Type message (Similar to the Constructor message user has provided, except, every native type become ValueType. Example 'int priority' becomes 'ValueType priority')
I have deleted all the configuration that we originally had in the Type message provided by the template author. I am assuming all of the things that was in our old Type message will now be part of adapter.params. May be some of the deleted configuration from template might still belong in the Constructor message. PTAL.
